e2e_tests/tests/test_system.py

```diff
@@ -595,6 +595,251 @@ def test_disable_and_enable_slots() -> None:
+    then be terminated promptly.
+    """
+    slots = _fetch_slots()
+    assert len(slots) == 1
+    agent_id = slots[0]["agent_id"]
+
+    command_id = _run_zero_slot_command(sleep=60)
+    # Wait for it to run.
+    _wait_for_command_state(command_id, "RUNNING", 30)
+
+    command = ["det", "-m", conf.make_master_url(), "agent", "disable", agent_id]
+    subprocess.check_call(command)
+    try:
+        _wait_for_command_state(command_id, "TERMINATED", 5)
+    finally:
+        command = ["det", "-m", conf.make_master_url(), "agent", "enable", agent_id]
+        subprocess.check_call(command)
+        # Kill the command before failing so it does not linger.
+        command = ["det", "-m", conf.make_master_url(), "command", "kill", command_id]
+        subprocess.check_call(command)
+
+
+@pytest.mark.e2e_cpu  # type: ignore
+def test_drain_agent() -> None:
+    """
+    Start an experiment, `disable --drain` the agent once the trial is running,
+    make sure the experiment still finishes, but the new ones won't schedule.
+    """
+
+    slots = _fetch_slots()
+    assert len(slots) == 1
+    agent_id = slots[0]["agent_id"]
+
+    experiment_id = exp.create_experiment(
+        conf.fixtures_path("no_op/single-medium-train-step.yaml"),
+        conf.fixtures_path("no_op"),
+        None,
+    )
+    exp.wait_for_experiment_state(experiment_id, "ACTIVE")
+    exp.wait_for_experiment_active_workload(experiment_id)
+    exp.wait_for_experiment_workload_progress(experiment_id)
+
+    command = ["det", "-m", conf.make_master_url(), "agent", "disable", "--drain", agent_id]
+    subprocess.check_call(command)
+    # Quickly enable it back.
+    command = ["det", "-m", conf.make_master_url(), "agent", "enable", agent_id]
+    subprocess.check_call(command)
+
+    # Try to launch another experiment. It shouldn't get scheduled because the
+    # slot is still busy with the first experiment.
+    experiment_id_no_start = exp.create_experiment(
+        conf.fixtures_path("no_op/single-medium-train-step.yaml"),
+        conf.fixtures_path("no_op"),
+        None,
+    )
+    time.sleep(5)
+    exp.wait_for_experiment_state(experiment_id_no_start, "ACTIVE")
+
+    command = ["det", "-m", conf.make_master_url(), "agent", "disable", "--drain", agent_id]
+    subprocess.check_call(command)
+
+    # Check for 15 seconds it doesn't get scheduled into the same slot.
+    for _ in range(15):
+        trials = exp.experiment_trials(experiment_id_no_start)
+        assert len(trials) == 0
+
+    # Ensure the first one has finished with the correct number of steps.
+    exp.wait_for_experiment_state(experiment_id, "COMPLETED")
+    trials = exp.experiment_trials(experiment_id)
+    assert len(trials) == 1
+    assert len(trials[0]["steps"]) == 5
+
+    # Ensure the slot is empty.
+    slots = _fetch_slots()
+    assert len(slots) == 1
+    assert slots[0]["enabled"] is False
+    assert slots[0]["draining"] is True
```

Comment:
I suppose I see `draining` more as a configuration option of the thing rather than the current state. so we could do that but it's extra complexity, so I'd skip that if you don't mind.

added the agent state check.